### PR TITLE
fix(python): profile every batch of a chunked Table, and accept any Arrow PyCapsule producer

### PR DESCRIPTION
**Pull request by AndreaBozzo**
## 📝 Summary

Two defects in `import_from_pyarrow`, found while exercising dataprof's Arrow
import path through the C Data Interface. They are independent and the commits
are separable — the first is a correctness bug, the second a documented-but-
missing capability.

**1. A chunked `Table` was profiled over its first batch only.**
`to_batches()` was followed by `batch_list[0]`, discarding the rest. A 12-row
Table in three batches reported 4 rows; the same 12 rows in one batch reported
12. Nothing in the report indicated the answer covered a prefix of the data.

**2. Only pyarrow's own types were accepted.** `profile()` documents accepting
"Arrow PyCapsule-compatible objects" and the Python layer routes anything with
`__arrow_c_array__` to `_profile_arrow`, but the import matched on the Python
type name being `Table`/`RecordBatch` and refused everything else. The
duck-typed branch exists in `convert_dataframe_to_batch` and is unreachable from
this path.

## 🔧 Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)

## 📋 Changes Made

- Import **every** batch of a `Table` and concatenate; the single-chunk case
  still avoids the copy.
- Fall through to `import_via_pycapsule` for any object exposing
  `__arrow_c_array__`. The pyarrow type names are still matched first, only so
  `Table` reaches its multi-batch handling.
- Reject non-struct arrays with a `TypeError` instead of panicking. This one is
  load-bearing: broadening what reaches the import path means a bare
  `pa.array([1,2,3])` — which *does* export a capsule — now gets there, and a
  record batch is specifically a *struct* array, so `StructArray::from` would
  have panicked with "entered unreachable code" across the FFI boundary. pyo3
  derives `PanicException` from `BaseException`, so callers catching `Exception`
  would not have caught it. Without this guard the change would have been worse
  than the behaviour it replaces.

## 🧪 Testing

New `python/tests/test_arrow_pycapsule_import.py`, 7 tests.

Confirmed the guards actually discriminate rather than passing vacuously:

| Build | Result |
|---|---|
| `master` (unfixed) | **4 failed**, 3 passed |
| with this branch | 7 passed |

The 3 that pass on master are the preservation checks — single-batch `Table`,
`RecordBatch`, and the plain-object refusal — so the fix is verified in the
state it must preserve as well as the state it improves. The non-struct test
discriminates against the intermediate state where change 2 was applied without
the struct guard, which panicked.

Full Python suite on this branch: **989 passed, 6 skipped** (skips are
uncompiled database features).

Before / after, same inputs:

```
                                    master        this branch
multi-batch Table (12 rows)         rows=4        rows=12
single-batch Table (4 rows)         rows=4        rows=4
RecordBatch                         rows=4        rows=4
PyCapsule producer (non-pyarrow)    TypeError     rows=4
bare pa.array (non-struct)          TypeError     TypeError
```

## Notes

Found while building an adversarial harness for the C Data / C Stream
Interface, which produces Arrow structures directly rather than through
pyarrow's constructors. There is a third, separate observation not addressed
here: an out-of-range dictionary index panics inside arrow-rs's `value()`
accessor, reached from `record_batch_analyzer.rs:107` via `ArrayFormatter`,
because `arrow::ffi::from_ffi` does not run full validation. That one needs a
design decision about where to pay for validation, so it is worth its own issue
rather than being folded in here.

### CODE REVIEW

File: crates/dataprof-python/src/arrow_export.rs [hunk 1]
```diff
@@ -889,25 +890,43 @@ fn convert_polars_to_batch(py: Python<'_>, df: &Bound<'_, PyAny>) -> PyResult<Re
...
+        let schema = imported[0].schema();
+        concat_batches(&schema, &imported).map_err(|e| {
+            PyRuntimeError::new_err(format!("Failed to combine Table batches: {}", e))
+        })
     } else if type_name == "RecordBatch" {
         import_via_pycapsule(py, obj)
+    } else if obj.hasattr("__arrow_c_array__")? {
+        // Any Arrow PyCapsule producer, which is what profile() documents
+        // accepting. pyarrow's own types are matched above only so that Table
+        // reaches its multi-batch handling; everything else imports the same way.
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply crates/dataprof-python/src/arrow_export.rs]: Fixed in bb61efd. The `Table` branch is now gated on the Table API rather than the type name alone:

```rust
if type_name == "Table" && obj.hasattr("to_batches")? {
```

so a PyCapsule producer whose class happens to be named `Table` falls through to the capsule import instead of failing on the missing `to_batches()`. Went with the API check rather than module/type identity because it does not assume the producer is pyarrow at all, which is the point of the contract.

File: python/tests/test_arrow_pycapsule_import.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    Stronger than the row count alone: it pins the property rather than one
+    number, so a future change that drops chunks in a different way still trips.
+    """
+    batch = _batch([1, 2, 3, 4])
+    chunked = pa.Table.from_batches([batch, batch, batch])
+    combined = chunked.combine_chunks()
+
+    assert dataprof.profile(chunked, name="chunked").rows == (
+        dataprof.profile(combined, name="combined").rows
+    )
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply python/tests/test_arrow_pycapsule_import.py]: Fair — that test did not guard what its name claimed. Rewritten in bb61efd.

It now uses distinct values per batch (`[1,2,3,4]`, `[10,20,30,40]`, `[100,200,300,400]`) so the statistics depend on every chunk, and compares `to_dict()["columns"]` rather than a row count. It also asserts the chunked result **differs** from profiling the first batch alone — without that, the equality would hold just as well for code that reads only the first chunk.

Also added a parametrized row-limit test for the new `limit_batches` path. Its boundary case is `max_rows=4` on 4-row batches: the limit is hit exactly with two batches still unread, which is the case a row-count comparison cannot classify.

File: crates/dataprof-python/src/arrow_export.rs [hunk 3]
```diff
@@ -889,25 +890,43 @@ fn convert_polars_to_batch(py: Python<'_>, df: &Bound<'_, PyAny>) -> PyResult<Re
...
+            imported.push(import_via_pycapsule(py, batch.bind(py))?);
+        }
+        if imported.len() == 1 {
+            // The common single-chunk case stays copy-free.
+            return Ok(imported.remove(0));
+        }
+        let schema = imported[0].schema();
+        concat_batches(&schema, &imported).map_err(|e| {
+            PyRuntimeError::new_err(format!("Failed to combine Table batches: {}", e))
+        })
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply crates/dataprof-python/src/arrow_export.rs]: Agreed, and taken the incremental route in bb61efd rather than working around it — `concat_batches` is gone from the profiling path entirely.

`import_batches_from_pyarrow` now returns the sequence and `profile_arrow` walks it with repeated `process_batch`, so a large chunked Table never has to exist as one contiguous RecordBatch. `import_from_pyarrow` survives as a thin concat wrapper only for `convert_dataframe_to_batch`, which cannot consume a sequence.

`limit_batches` applies the row limit across the sequence and stops early, so a bounded profile no longer imports rows it will discard. The subtle part is the truncation flag: it has to be decided while walking, not by comparing a final row count against the limit, because stopping at exactly `limit` rows with batches still unread *is* truncation and a count alone cannot see that. A zero limit keeps an empty slice so the schema, and therefore the column count, still reaches the report — matching what `limit_batch_rows` did via `slice(0, 0)`.

One thing worth being straight about: **no test in this PR distinguishes incremental processing from concat-then-slice.** Both produce identical output; the difference is peak memory and the >2 GiB offset overflow you flagged, neither of which the suite observes. The parametrized row-limit test guards the *semantics* of the limit (including the batch-boundary case), not the allocation behaviour. If you want the memory property pinned, that needs a separate benchmark or a large-input test, which felt out of scope here.
